test/distributed/_composable/test_composability/test_pp_composability.py

```diff
@@ -357,6 +377,172 @@ def _dcp_test(self):
 
         _dcp_test(self)
 
+    @requires_nccl()
+    @skip_if_lt_x_gpu(8)
+    @skip_but_pass_in_sandcastle_if(not TEST_MULTIGPU, "Test requires 8+ GPUs")
+    @parametrize(
+        "ScheduleClass",
+        [
+            ScheduleGPipe,
+            Schedule1F1B,
+            ScheduleInterleaved1F1B,
+            ScheduleLoopedBFS,
+            ScheduleInterleavedZeroBubble,
+        ],
+    )
+    def test_3d_with_tp_dp_pp(self, ScheduleClass):
+        device = torch.device("cuda", self.device)
+        torch.cuda.set_device(self.device)
+        store = torch.distributed.FileStore(self.file_name, self.world_size)
+        torch.distributed.init_process_group(
+            backend="nccl",
+            store=store,
+            rank=self.rank,
+            world_size=self.world_size,
+        )
+        tp_size = 2
+        pp_size = 2
+        num_microbatches = 8
+        dp_size = self.world_size // (tp_size * pp_size)
+        device_mesh = init_device_mesh(
+            "cuda",
+            mesh_shape=(dp_size, pp_size, tp_size),
+            mesh_dim_names=("dp", "pp", "tp"),
+        )
+        dp_mesh = device_mesh["dp"]
+        tp_mesh = device_mesh["tp"]
+        pp_mesh = device_mesh["pp"]
+        pp_group = device_mesh["pp"].get_group()
+
+        # create "entire model"
+        total_layers = 8
+        dim = 8
+        full_model = nn.ModuleList([MLPModuleEven() for _ in range(total_layers)])
+        ref_model = nn.Sequential(*copy.deepcopy(full_model))
+        ref_model.to(self.device)
+
+        # dummy loss needed just to force backwards to run in schedule step
+        def loss_fn(y, target):
+            return y.sum()
+
+        # Apply DP to stage module
+        def apply_dp(partial_model):
+            # apply FSDP
+            mp_policy = MixedPrecisionPolicy(
+                param_dtype=torch.float32,
+                reduce_dtype=torch.float32,
+            )
+            fsdp_config = {"mesh": dp_mesh, "mp_policy": mp_policy}
+            for layer_id in range(len(partial_model)):
+                fully_shard(
+                    partial_model[layer_id],
+                    **fsdp_config,
+                    reshard_after_forward=False,
+                )
+            dp_model = fully_shard(partial_model, **fsdp_config)
+            return dp_model
+
+        def apply_tp(
+            model: nn.Module,
+            tp_mesh: DeviceMesh,
+        ):
+            parallelize_plan = {
+                "net1": ColwiseParallel(),
+                "net2": RowwiseParallel(),
+                "net3": ColwiseParallel(),
+            }
+            for layer in model:
+                parallelize_module(layer, tp_mesh, parallelize_plan)
+            return model
+
+        # Attach to a schedule
+        if issubclass(ScheduleClass, PipelineScheduleSingle):
```

Comment:
note: not in this PR, but we might want to clean this pp init stuff up later.  we should get rid of the 'single' vs 'multi' base classes and make it easier to construct a schedule without so many lines of boilerplate.  Or maybe i'm wrong and single vs multi is legitimately worth having different init flows for?? cc @H-Huang